### PR TITLE
BTS-1868: fix TSan errors

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Backport of https://github.com/arangodb/arangodb/pull/20937
Enterprise companion PR: https://github.com/arangodb/enterprise/pull/1485

BTS-1868: fix TSan errors
https://arangodb.atlassian.net/browse/BTS-1868

This PR makes 2 relevant changes:
* NetworkFeature now shuts down retry requests also in its destructor. Previously retry requests were only shut down in `stop()` and `unprepare()`, but if these were not called, retry requests could leak.
* `AuthenticationFeature::INSTANCE` was a raw pointer and was accessed and modified without synchronization. This PR turns it into an atomic pointer, so that TSan does not report data races when the member is accessed.

As this PR also removes some includes from central header files and replaces them by forward definitions, a lot of unrelated files also had to be changed so that the previously transitively included header files would be included there.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.12.0: -
  - [ ] Backport for 3.11: this PR
  - [ ] Backport for 3.10: -

#### Related Information

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1485
- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/BTS-1868
- [ ] Design document:

### CODE REVIEW

**Review by neunhoef — APPROVED** [review]
LGTM

File: arangod/GeneralServer/AuthenticationFeature.cpp [hunk 1]
```diff
@@ -302,7 +306,7 @@ void AuthenticationFeature::prepare() {
   _authCache->setJwtSecret(_jwtSecretProgramOption);
 #endif
 
-  INSTANCE = this;
+  INSTANCE.store(this, std::memory_order_relaxed);
```
Comment (mpoeter) [inline arangod/GeneralServer/AuthenticationFeature.cpp:309]:
```suggestion
  INSTANCE.store(this, std::memory_order_release);
```

File: arangod/GeneralServer/AuthenticationFeature.cpp [hunk 2]
```diff
@@ -323,7 +327,37 @@ void AuthenticationFeature::start() {
   LOG_TOPIC("3844e", INFO, arangodb::Logger::AUTHENTICATION) << out.str();
 }
 
-void AuthenticationFeature::unprepare() { INSTANCE = nullptr; }
+void AuthenticationFeature::unprepare() {
+  INSTANCE.store(nullptr, std::memory_order_relaxed);
+}
+
+AuthenticationFeature* AuthenticationFeature::instance() noexcept {
+  return INSTANCE.load(std::memory_order_relaxed);
```
Comment (mpoeter) [inline arangod/GeneralServer/AuthenticationFeature.cpp:335]:
```suggestion
  return INSTANCE.load(std::memory_order_acquire);
```

**Review by mpoeter — CHANGES_REQUESTED** [review]

**Comment by mpoeter** [comment]
Two small issues, otherwise LGTM

**Review by mpoeter — APPROVED** [review]

**Pull request merged**